data-prepper-event/src/test/java/org/opensearch/dataprepper/core/event/CachingEventKeyFactoryTest.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.core.event;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.opensearch.dataprepper.model.event.EventKey;
+import org.opensearch.dataprepper.model.event.EventKeyFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import static org.hamcrest.CoreMatchers.sameInstance;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.atLeast;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class CachingEventKeyFactoryTest {
+    private static final int CACHE_SIZE = 2;
+    @Mock
+    private EventKeyFactory innerEventKeyFactory;
+
+    @Mock
+    private EventConfiguration eventConfiguration;
+
+    @BeforeEach
+    void setUp() {
+        when(eventConfiguration.getMaximumCachedKeys()).thenReturn(CACHE_SIZE);
+    }
+
+    private EventKeyFactory createObjectUnderTest() {
+        return new CachingEventKeyFactory(innerEventKeyFactory, eventConfiguration);
+    }
+
+    @ParameterizedTest
+    @EnumSource(EventKeyFactory.EventAction.class)
+    void createEventKey_with_EventAction_returns_inner_createEventKey(final EventKeyFactory.EventAction eventAction) {
+        final String key = UUID.randomUUID().toString();
+        final EventKey eventKey = mock(EventKey.class);
+
+        when(innerEventKeyFactory.createEventKey(key, eventAction)).thenReturn(eventKey);
+
+        final EventKey actualEventKey = createObjectUnderTest().createEventKey(key, eventAction);
+        assertThat(actualEventKey, sameInstance(eventKey));
+    }
+
+    @Test
+    void createEventKey_returns_inner_createEventKey() {
+        final String key = UUID.randomUUID().toString();
+        final EventKey eventKey = mock(EventKey.class);
+
+        when(innerEventKeyFactory.createEventKey(key, EventKeyFactory.EventAction.ALL)).thenReturn(eventKey);
```

Comment:
What is this test testing? Since we are returning via `when` it will be same instance, right? Isn't it better to use the real `createEventKey()`?

Reply:
Overall, this test is verifying the behavior of `createEventKey(String)` in the `CachingEventKeyFactory`. The interface has a default implementation for `createEventKey(final String key)`:

https://github.com/opensearch-project/data-prepper/blob/8eef2f68a8c54ed2aee8eb50e03fcf7fcc784d4b/data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/EventKeyFactory.java#L37-L39

The `CachingEventKeyFactory` actually uses this default implementation, which then calls the custom implementation for caching.

The real implementations are being used in this test. Please note that I am not using a spy on the object under test. On this line, I have mocked the _inner_ object. This is the same as on line 55 above. The difference is that because of the default implementation, the call comes in providing `EventAction.ALL`.